common/src/main/java/org/apache/rocketmq/common/client/ClientManagerImpl.java

```diff
@@ -0,0 +1,217 @@
+                }
+            }
+            oldClient.setLastUpdateTimestamp(System.currentTimeMillis());
+            onRegister(groupId, remotingChannel);
+        }
+        log.debug("Register client role: {}, group: {}, last: {}", client.getClientRole(), groupId,
+            client.getLastUpdateTimestamp());
+        return updated;
+    }
+
+    protected void removeClient(String groupId, RemotingChannel remotingChannel) {
+        ConcurrentHashMap<RemotingChannel, Client> channelTable = groupClientTable.get(groupId);
+        if (channelTable != null) {
+            Client prev = channelTable.remove(remotingChannel);
+            if (prev != null) {
+                log.info("Unregister client: {}  in  group, {}", prev, groupId);
+            }
+            if (channelTable.isEmpty()) {
+                groupClientTable.remove(groupId);
+                log.info("Unregister client ok, no any connection, and remove consumer group, {}",
+                    groupId);
+            }
+        }
+    }
+
+    @Override
+    public void unRegister(String groupId, RemotingChannel remotingChannel) {
+        removeClient(groupId, remotingChannel);
+        onUnregister(groupId, remotingChannel);
+
+    }
+
+    @Override
+    public void onClose(Set<String> groups, RemotingChannel remotingChannel) {
+        for (String groupId : groups) {
+            removeClient(groupId, remotingChannel);
+            onClosed(groupId, remotingChannel);
+        }
+    }
+
+    public List<RemotingChannel> getChannels(String groupId) {
+        if (groupId != null) {
+            List<RemotingChannel> result = new ArrayList<RemotingChannel>();
+            ConcurrentHashMap channelsMap = this.groupClientTable.get(groupId);
+            if (channelsMap != null) {
+                result.addAll(this.groupClientTable.get(groupId).keySet());
+                return result;
+            }
+            return null;
+        }
+        return null;
+    }
+
+    @Override
+    public List<String> getAllClientId(String groupId) {
+        List<String> result = new ArrayList<String>();
+        Map<RemotingChannel, Client> channelClientMap = this.groupClientTable.get(groupId);
+        if (channelClientMap != null) {
+            Iterator<Map.Entry<RemotingChannel, Client>> it = channelClientMap.entrySet()
+                .iterator();
+            while (it.hasNext()) {
+                Map.Entry<RemotingChannel, Client> entry = it.next();
+                Client client = entry.getValue();
+                result.add(client.getClientId());
+            }
+        }
+        return result;
+    }
+
+    @Override
+    public Client getClient(String groupId, RemotingChannel remotingChannel) {
+        assert groupId != null && remotingChannel != null;
+        if (!groupClientTable.containsKey(groupId)) {
+            return null;
+        }
+        ConcurrentHashMap<RemotingChannel, Client> channelClientMap = groupClientTable
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **207 and 210**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The remove that can remove the key is at line: **153**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).